### ISSUE TITLE
`SUBSCRIBE` envelope upsert

### ISSUE THREAD

**Issue by awang** [MaterializeInc/materialize#5875]
*Note from @benesch.*

We'd likely implement this as an optional `ENVELOPE UPSERT` clause, or somesuch, that you attach to the end of a `SUBSCRIBE` statement.

## Requirements 

- Add option to SUBSCRIBE, gated behind a feature flag: `ENVELOPE UPSERT (KEY <expr>)`
- When specified, the output of the subscribe becomes (mz_timestamp, mz_key, mz_value). mz_key is the specified key expression and mz_value is the data. There is no diff field.
- Within each timestamp, the upsert envelope accumulates updates into a series of diff pairs (i.e., before/after pairs). Diff pairs representing an insert or update are emitted with mz_value set to the after value. Diff pairs representing a delete are emitted with an mz_value of NULL
- The type of mz_key is the type of the key expression. The type of mz_value is a record type matching the row type of the SUBSCRIBE query.
- Must be compatible with existing WITH options (snapshot, progress)

## Example
Using [get started](https://materialize.com/docs/get-started/). View:
```
SELECT item_name, item_category, revenue FROM item_summary_5min LIMIT 5;
 item_id |     item_name      | item_category |  revenue
---------+--------------------+---------------+-----------
    1024 | health spark       | gadgets       | 514461.99
     512 | vision liquid      | widgets       | 158461.92
     513 | department poppy   | doodads       |    220248
    1025 | kangaroo bomber    | gadgets       |  286929.6
    1792 | Linda transmission | widgets       | 211137.28
```

```
-- Normal output (no envelope)
-- For these examples assume this list represents a full set of updates, e.g. +1/-1 is an update, while a lone -1 is a delete.
COPY (
    SUBSCRIBE (
        SELECT item_id, item_name, item_category, revenue
        FROM item_summary_5min
)) TO STDOUT;

1680012250628	1	706	relatives anethesiologist	clearance	493301.48
1680012251000	1	313	title kite	widgets	310590
1680012251000	-1	313	title kite	widgets	310037.84
1680012251000	-1	44	jumbo Malaysia	doodads	52358.4
1680012251000	-1	899	case hardhat	gadgets	343678.8
1680012251000	1	899	case hardhat	gadgets	345165.3
```

```
COPY (
    SUBSCRIBE (
        SELECT item_name, item_category, revenue
        FROM item_summary_5min
) ENVELOPE UPSERT (KEY item_id)) TO STDOUT;

- mz_timestamp  mz_key  mz_value
1680012250628   706     relatives anethesiologist   clearance   493301.48
1680012251000   313     title kite                  widgets     310590
1680012251000   44      NULL
1680012251000   899     case hardhat                gadgets     345165.3
```
```
-- With progress enabled 
COPY (
    SUBSCRIBE (
        SELECT item_name, item_category, revenue
        FROM item_summary_5min
) ENVELOPE UPSERT (KEY item_id) WITH (progress = true)) TO STDOUT;

-- mz_timestamp progress    mz_key  mz_value
1680012250628   F           706     relatives anethesiologist   clearance   493301.48
1680012251000   F           313     title kite                  widgets     310590
1680012251000   F           44      NULL
1680012251000   F           899     case hardhat                gadgets     345165.3
1680012251001   T           NULL    NULL
```
-----

*Original issue.*

Tracking issue to discuss a proposal that came out of last week's meeting. Some internal discussion [already happened here](https://materializeinc.slack.com/archives/CUA4B6X55/p1613578154168200):

A user of TAIL currently has to maintain a hash table to generate a cdc-style output stream from TAIL.  Moving to a key/value format would make things easier to match than the +1/-1 format we use today.  

```
Basically they want to write the following logic (in pseudocode that looks a lot like Rust):
let mut current_val_of_UK = None;
while let Some(next) = tail.next() {
    if next.key == "UK" {
        if next.diff > 0 {
            current_val_of_UK = Some(next);
        } else if next.diff < 0 {
            current_val_of_UK = None;
        }
    }
}

If the `-1`s were ordered before the `+1`s, this logic will result in current_val_of_UK being correct, assuming all the messages for the timestamp have been processed

```

While putting deletes before inserts will make it easier to write consumers of TAIL, one worry is that it will become a footgun for people. Users may not realize that waiting for progress messages is important.

**Comment by krishmanoh2** [comment]
Could we generate similar output as avro key/value upsert similar to sink as an option?

**Comment by ruchirK** [comment]
> Could we generate similar output as avro key/value upsert similar to sink as an option?

If I understand this right we could get upsert style semantics if we suppressed retractions.

But also, I just want to say again, that they can make things work in their current setup by just ignoring retractions that don't match the value they are currently holding.

as an example: if they currently have `(foo, bar)` in their data set and then they see

```
((foo, baz), +1)
((foo, bar) -1)
```

once they've applied `(foo, baz)` they just need to ignore the `(foo, bar)` retraction because the value doesn't match to make things work ok. In brennans pseudocode this would be:
```
Basically they want to write the following logic (in pseudocode that looks a lot like Rust):
let mut current_val_of_UK = None;
while let Some(next) = tail.next() {
    if next.key == "UK" {
        if next.diff > 0 {
            current_val_of_UK = Some(next);
        } else if next.diff < 0 && next.val == current_val_of_UK { // this is the only line that changed. only apply retractions on matching values! 
            current_val_of_UK = None;
        }
    }
}
```
Can we see if users are able to try this first?

**Comment by cirego** [comment]
How would we tell Materialize which part of the result should be a key and which should be a value? Do we have a proposal for how to remap tuples into key/value pairs?
